core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1644,23 +1644,32 @@ object Stream {
     def concurrently[O2](that: Stream[F, O2])(implicit F: Effect[F],
                                               ec: ExecutionContext): Stream[F, O] =
       Stream.eval(async.promise[F, Unit]).flatMap { interruptR =>
-        Stream.eval(async.promise[F, Unit]).flatMap { doneR =>
+        Stream.eval(async.promise[F, Option[Throwable]]).flatMap { doneR =>
           Stream.eval(async.promise[F, Throwable]).flatMap { interruptL =>
             def runR =
               that
                 .interruptWhen(interruptR.get.attempt)
                 .compile
                 .drain
                 .attempt
-                .flatMap {
-                  case Right(_)  => doneR.complete(())
-                  case Left(err) => interruptL.complete(err) *> doneR.complete(())
+                .map { _.left.toOption }
+                .flatMap { r =>
+                  doneR.complete(r) >> // to prevent deadlock, done must be signalled before `interruptL`
```

Comment:
I'm having some difficulty understanding why this is the case. Let's say you have the opposite:
`r.fold(F.unit)(interruptL.complete) >> doneR.complete(r)`

could you describe a scenario when this causes a deadlock?

Reply:
It only manifests itself when failing with exception. Think the situation when the `R` fails, but not yet signals its termination (doneR is after interruptL), then, with that failure, the `L` side is interrupted and in its onFinalize code, it awaits the `R` to complete (doneR.get) (in new implementation this is in append). That is indeed deadlock. 
Also note that failure may happen immediately `after` the `L` completes, and w/o this approach failure failed to be propagated to `L` stream in such case. 

Not sure if it is clear with explanation, but given the fact it took me like 3 days to find whats going on, I am quite sure it is not very `transparent` issue.